e2e: opertor's standard mode

jiayouxujin opened the pull request:
Hi,
This PR is for e2e-test.

jiayouxujin commented on test/e2e/standard/standard_test.go:
```diff
@@ -195,47 +246,52 @@ var _ = Describe("KVRocks standard controller", func() {
 		}
-		if err := suite.K8sClient.Get(suite.CTX, key, &pod); err != nil {
+		// portforward for e2e test
+		localPort := []string{fmt.Sprintf("%d:%d", 6379+index, 6379)}
```
@tianshimoyi Hi, if I want to connect to kvrocks in e2e tests, are there any other elegant methods besides portforward?

jiayouxujin replied:
I will remove that by using the telepresence tool.

ColinChamber commented on test/e2e/README.md:
````diff
@@ -0,0 +1,80 @@
+
+```go
+var config = &Config{
+	InstallKubernetes:        false,
````
`InstallKubernetes` might not be clear enough, it could easily confuse users. What we're doing here is creating a k8s cluster named 'e2e-test' within minikube. I would also suggest replacing 'InstallKubernetes' in other places where it's used.

jiayouxujin replied:
@ColinChamber Hi，
I have removed this value and now determine whether to install Kubernetes based on whether the `kubeconfig` parameter is empty or not.

ColinChamber commented on test/e2e/standard/standard_test.go:
```diff
@@ -1,53 +1,95 @@
 )
 
-var _ = Describe("KVRocks standard controller", func() {
+// TODO remove the hard code and use the config.yaml
```
I believe that currently 'Cluster setup' and 'Only code' are essentially the same, both creating a cluster through minikube for testing. If we wanted to provide two options, they should be creating a cluster and using an existing cluster. Of course, the config.yaml is in a 'to-do' state.

jiayouxujin replied:
@ColinChamber Hi,
Now we can use the `config.yaml` file.

git-hulk commented on test/e2e/cluster/cluster.go:
```diff
@@ -1,5 +1,28 @@
+// 	kvrocksv1alpha1 "github.com/RocksLabs/kvrocks-operator/api/v1alpha1"
+// 	"github.com/RocksLabs/kvrocks-operator/pkg/client/kvrocks"
+// 	"github.com/RocksLabs/kvrocks-operator/pkg/resources"
+// 	"github.com/RocksLabs/kvrocks-operator/test/e2e/suite"
```
@jiayouxujin We should remove the code directly since we're using the version control service and can take it back if we want.

jiayouxujin replied:
Ok, I will remove the unnecessary code.

jiayouxujin committed:
rebaes

jiayouxujin committed:
remove code

remove println

add readme

remove the portForward

remove the UsingKvorcksOperatorHelm

fix: fix the error of certificate

feat: introduce config.yaml

update default config.yaml

remove the unnecessary code

add telepresence for default config

jiayouxujin committed:
add pvc check

jiayouxujin commented on test/e2e/standard/standard_test.go:
```diff
@@ -310,6 +311,12 @@ func checkKVRocks(instance *kvrocksv1alpha1.KVRocks) error {
 	}
 
-	// TODO check PVC
+	var pvcList corev1.PersistentVolumeClaimList
```
Now, pvc checker is supported.

ColinChamber commented on test/e2e/README.md:
````diff
@@ -0,0 +1,40 @@
+```yaml
+kruiseVersion: 1.4.0
+clusterName: e2e-test
+namespace: kvrocks
````
The namespace cannot be changed to another value.

jiayouxujin replied:
Since the default namespace provided in the [yaml file is `kvrocks`](https://github.com/RocksLabs/kvrocks-operator/blob/unstable/examples/standard.yaml#L5), if you want to modify the namespace here, you need to update the configuration in the yaml file as well. I have already added this information to the readme.

ColinChamber commented on test/e2e/README.md:
```diff
@@ -0,0 +1,40 @@
+kruiseVersion: 1.4.0
+clusterName: e2e-test
+namespace: kvrocks
+kubeConfig: /path/to/your/kubeconfig
```
After setting kubeconfig, is it still necessary to fill in the clusterName?

jiayouxujin replied:
Thanks, it has been added to the readme.

ColinChamber commented on test/e2e/util/config.go:
```diff
@@ -0,0 +1,45 @@
+	KruiseVersion string `yaml:"kruiseVersion"`
+	ClusterName   string `yaml:"clusterName"`
+	Namespace     string `yaml:"namespace"`
+	KubeConfig    string `yaml:"kubeConfig"`
```
Do all variables declared in a struct need to be exported?

jiayouxujin replied:
That's helpful. I have already hidden some variables.

ColinChamber replied:
Sorry, I overlooked that the members of Config need to be exported in order to use `yaml.Unmarshal(configData, config).`

jiayouxujin replied:
I have fixed the bug.

ColinChamber commented on test/e2e/util/kubernetes_env.go:
```diff
@@ -0,0 +1,175 @@
+
+	// Add OpenKruise Helm repo
+	repoList, err := HelmTool("repo", "list")
+	Expect(err).NotTo(HaveOccurred())
```
Tolerate the 'no repositories to show' error.

jiayouxujin replied:
Thanks, I have resolved this bug.

jiayouxujin committed:
fix: fix helm repo error

jiayouxujin committed:
fix: update the variable

jiayouxujin committed:
fix: fix the minikube script

jiayouxujin committed:
update the readme

ColinChamber commented on test/e2e/util/kubernetes_env.go:
```diff
@@ -0,0 +1,178 @@
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	if !env.isHelmInstalled("kvrocks-operator") {
```
There is a slight issue here. When using helm status to check if kvrocks-operator is installed, you need to specify the namespace.

ColinChamber replied:
```sh
sh-4.2$ helm status kvrocks-operator
Error: release: not found
sh-4.2$ helm status kvrocks-operator -n kvrocks
NAME: kvrocks-operator
LAST DEPLOYED: Thu Jun 29 14:55:02 2023
NAMESPACE: kvrocks
STATUS: deployed
REVISION: 1
TEST SUITE: None

jiayouxujin replied:
Thanks, I have fixed the bug.

jiayouxujin committed:
fix: fix the vals of config

jiayouxujin committed:
fix: isHelmInstalled

ColinChamber committed:
fix: check ns

ColinChamber committed:
add: check clusterName

ColinChamber committed:
fix: use parseManifest

jiayouxujin committed:
fix: instance is nil

jiayouxujin committed:
update readme

ColinChamber committed:
typo: lowercase manifestDir

ColinChamber reviewed (approved):
A Terrific job! Thanks @jiayouxujin
